core/sparse_array.cpp

```diff
@@ -46,6 +47,25 @@ SparseArray::SparseArray(unsigned int p_nrows, unsigned int p_ncols)
 	
 	row_offsets_[nrows_set_] = 0;
 	finished_ = false;
+
+    //Allocate space for 2-D vectors
+    //Columns and Distances are now 2-D vectors with each row corresponding to one focal individual. This negates the need for row offsets.
+
+    initial_width=5000;       //Some random initial width for each row. Can be changed later.
+    columns = (uint32_t **)malloc((nrows_) * sizeof(uint32_t *));
+    distances = (uint32_t **)malloc((nrows_) * sizeof(uint32_t *));
+    nnz = (uint32_t *)malloc((nrows_) * sizeof(uint32_t));
+    nnz_capacity = (uint32_t *)malloc((nrows_) * sizeof(uint32_t));
+
+    for(int i=0;i<nrows_;i++)
+    {   
+        columns[i]=(uint32_t *)malloc((initial_width) * sizeof(uint32_t));
+        distances[i]=(uint32_t *)malloc((initial_width) * sizeof(uint32_t));
+    
+    //initial nnz and its capacity
+        nnz[i]=0;
+        nnz_capacity[i]=0;
```

Comment:
this should be `initial_width`

Reply:
You know, rather than pre-allocating all the rows at some initial width, you could actually just leave columns[i], distances[i], etc. as nullptr, it seems to me, and have nnz_capacity[i] be zero as you have now.  That would have the effect of deferring allocation of the row until the point when the row is written to for the first time, and would allow that allocation to be done by the threads rather than in a single-threaded block up front.  It would also have the benefit that if a row remains empty (because a given individual does not interact with any other individual) the row's buffers never get allocated at all – a savings in time and memory for models where that is common.  I think that design will probably be both simpler (all allocation is done by realloc) and potentially more efficient.

Reply:
Thanks for that. In general, that looks like the better way of creating 2d vectors. I've always created arrays first to some initial usage and then worry about reallocating. A relatively bad habit that I would rectify.